### PR TITLE
IP restrictions

### PR DESCRIPTION
**Pull request by sroettger**
Implement IP restrictions for ingress + load balancer

The Ingress backend will now have a backend-config that points to a cloud armor security policy. In this policy, we can filter based on IP addresses.

Similarly, the LoadBalancer service now uses the LoadBalancerSourceRanges option to filter based on source IP.

The source of truth is stored in the cloud armor policy rule with priority 0. It will get applied automatically to the load balancers after a few minutes.
The operator is passed the ranges in an environment variable during startup. In order to update it, we restart the operator.

### CODE REVIEW

File: dist/bin/kctf-cluster [hunk 1]
```diff
@@ -320,11 +320,98 @@ EOF
...
+
+  RANGES="$1"
+
+  # update the cloud armor policy
+  set_cloud_armor_policy "${RANGES}" || return
+
+  # stop the operator
+  "${KCTF_BIN}/kubectl" delete -f "${KCTF_CTF_DIR}/kctf/resources/operator.yaml" || return
+
+  start_operator_gce || return
```
Comment (sirdarckcat) [inline dist/bin/kctf-cluster:386]: what happens when someone runs this on a local cluster?

File: dist/bin/kctf-cluster [hunk 2]
```diff
@@ -320,11 +320,98 @@ EOF
...
+
+function start_operator_gce {
+  get_cloud_armor_policy || return
+  RANGES=$ret
+  # restart the operator with the new range
+  "${KCTF_BIN}/yq" eval "(.spec.template.spec.containers[0].env[] | select(.name == \"ALLOWED_IPS\")).value = \"${RANGES}\"" "${KCTF_CTF_DIR}/kctf/resources/operator.yaml" | "${KCTF_BIN}/kubectl" apply -f - || return
+}
+
+function set_cloud_armor_policy {
+  gcloud compute security-policies rules update --action allow --security-policy kctf-policy --src-ip-ranges "$1" 0 || return
```
Comment (sirdarckcat) [inline dist/bin/kctf-cluster]: this probably needs to be namespaced (to support more than 1 cluster per project) like the other kctf stuff

File: dist/bin/kctf-cluster [hunk 3]
```diff
@@ -320,11 +320,98 @@ EOF
...
+  ret=$(gcloud compute security-policies rules describe --security-policy kctf-policy 0 --format 'value[delimiter=","](match.config.srcIpRanges)')
+}
+
+function create_cloud_armor_policy {
+  if [[ -z $(gcloud compute security-policies list --filter "name=kctf-policy" --format 'get(name)') ]]; then
+    gcloud compute security-policies create kctf-policy || return
+    gcloud compute security-policies rules create --action allow --security-policy kctf-policy --src-ip-ranges "0.0.0.0/0" 0 || return
+
+    # make sure the default policy is deny
+    gcloud compute security-policies rules update --action deny-404 --security-policy kctf-policy 2147483647 || return
```
Comment (sirdarckcat) [inline dist/bin/kctf-cluster]: add a comment for whats the number?

File: dist/bin/kctf-cluster [hunk 4]
```diff
@@ -747,12 +836,14 @@ function kctf_cluster_resize {
...
-  echo -e "  stop:   stop the cluster" >&2
-  echo -e "  resize: resize the cluster" >&2
+  echo -e "  create:            create a new cluster config" >&2
+  echo -e "  list:              list available cluster configs" >&2
+  echo -e "  load:              load an existing cluster config" >&2
+  echo -e "  start:             start the cluster" >&2
+  echo -e "  stop:              stop the cluster" >&2
+  echo -e "  resize:            resize the cluster" >&2
+  echo -e "  set-src-ip-ranges: allow a range of IPs to access the services" >&2
+  echo -e "  get-src-ip-ranges: print the current allowed IP range" >&2
```
Comment (sirdarckcat) [inline dist/bin/kctf-cluster]: nit: plural in name singular in text. it's ranges I guess

File: dist/bin/kctf-cluster (hunk 2 above)
Comment (sroettger) [reply dist/bin/kctf-cluster]: I don't think we actually support more than one cluster per project and it makes stuff more complicated.
Do we actually want to support this?

File: dist/bin/kctf-cluster (hunk 1 above)
Comment (sroettger) [reply dist/bin/kctf-cluster:386]: added a check for type == gce

File: dist/bin/kctf-cluster (hunk 3 above)
Comment (sroettger) [reply dist/bin/kctf-cluster]: added a comment

File: dist/bin/kctf-cluster (hunk 4 above)
Comment (sroettger) [reply dist/bin/kctf-cluster]: done

File: dist/bin/kctf-cluster (hunk 2 above)
Comment (sroettger) [reply dist/bin/kctf-cluster]: in this case, we would need to pass the name to the operator for example

Comment (sirdarckcat) [reply dist/bin/kctf-cluster]: #133 fixed the issues we had with multi-cluster support. this means v1.1 doesn't support it (so it's kindof a breaking change)

Comment (sirdarckcat) [reply dist/bin/kctf-cluster]: oops, I meant #140

Comment (sroettger) [reply dist/bin/kctf-cluster]: that bug is still open though

Comment (sirdarckcat) [reply dist/bin/kctf-cluster]: Yes but we only have issues at the challenge name level. It's already possible to have 2 clusters with the same project, we just haven't made the challenge names namespaced too.

If we don't fix this here, then it'll be effectively impossible to have 2 clusters in one project (right now it's just not possible to have 2 tasks with the same name).

**Comment by sroettger** [comment]
Isn't the ip range still hardcoded too?

**Comment by sirdarckcat** [comment]
no, we have custom subnets now https://github.com/google/kctf/pull/208

**Comment by sirdarckcat** [comment]
err, we have custom networks now ;P

File: dist/bin/kctf-cluster (hunk 2 above)
Comment (sroettger) [reply dist/bin/kctf-cluster]: Using a unique policy name now

**Review by sirdarckcat — APPROVED** [review]

**Pull request merged**